airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/state/GlobalStateManager.java

```diff
@@ -0,0 +1,129 @@
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Global implementation of the {@link StateManager} interface.
+ *
+ * This implementation generates a single, global state object for the state tracked by this
+ * manager.
+ */
+public class GlobalStateManager extends AbstractStateManager<AirbyteStateMessage, AirbyteStreamState> {
+
+  /**
+   * Legacy {@link CdcStateManager} used to manage state for connectors that support Change Data
+   * Capture (CDC).
+   */
+  private final CdcStateManager cdcStateManager;
+
+  /**
+   * Constructs a new {@link GlobalStateManager} that is seeded with the provided
+   * {@link AirbyteStateMessage}.
+   *
+   * @param airbyteStateMessage The initial state represented as an {@link AirbyteStateMessage}.
+   * @param catalog The {@link ConfiguredAirbyteCatalog} for the connector associated with this state
+   *        manager.
+   */
+  public GlobalStateManager(final AirbyteStateMessage airbyteStateMessage, final ConfiguredAirbyteCatalog catalog) {
+    super(catalog,
+        getStreamsSupplier(airbyteStateMessage),
+        CURSOR_FUNCTION,
+        CURSOR_FIELD_FUNCTION,
+        NAME_NAMESPACE_PAIR_FUNCTION);
+
+    this.cdcStateManager = new CdcStateManager(extractCdcState(airbyteStateMessage));
+  }
+
+  @Override
+  public CdcStateManager getCdcStateManager() {
+    return cdcStateManager;
+  }
+
+  @Override
+  public AirbyteStateMessage toState(final AirbyteStreamNameNamespacePair pair) {
+    // Populate global state
+    final AirbyteGlobalState globalState = new AirbyteGlobalState();
+    globalState.setSharedState(Jsons.jsonNode(getCdcStateManager().getCdcState()));
+    globalState.setStreamStates(StateGeneratorUtils.generateStreamStateList(getPairToCursorInfoMap()));
+
+    // Generate the legacy state for backwards compatibility
+    final DbState dbState = StateGeneratorUtils.generateDbState(getPairToCursorInfoMap())
+        .withCdc(true)
+        .withCdcState(getCdcStateManager().getCdcState());
+
+    return new AirbyteStateMessage()
+        .withStateType(AirbyteStateType.GLOBAL)
+        // Temporarily include legacy state for backwards compatibility with the platform
+        .withData(Jsons.jsonNode(dbState))
+        .withGlobal(globalState);
+  }
+
+  /**
+   * Extracts the Change Data Capture (CDC) state stored in the initial state provided to this state
+   * manager.
+   *
+   * @param airbyteStateMessage The {@link AirbyteStateMessage} that contains the initial state
+   *        provided to the state manager.
+   * @return The {@link CdcState} stored in the state, if any. Note that this will not be {@code null}
+   *         but may be empty.
+   */
+  private CdcState extractCdcState(final AirbyteStateMessage airbyteStateMessage) {
+    if (airbyteStateMessage.getStateType() == AirbyteStateType.GLOBAL) {
+      return Jsons.object(airbyteStateMessage.getGlobal().getSharedState(), DbState.class).getCdcState();
```

Comment:
Here is my concern with this method, when we emit the state we split the JSON into `sharedState` and `streams`
```
  @Override
  public AirbyteStateMessage toState(final AirbyteStreamNameNamespacePair pair) {
    // Populate global state
    final AirbyteGlobalState globalState = new AirbyteGlobalState();
    globalState.setSharedState(Jsons.jsonNode(getCdcStateManager().getCdcState()));
    globalState.setStreamStates(StateGeneratorUtils.generateStreamStateList(getPairToCursorInfoMap()));

    // Generate the legacy state for backwards compatibility
    final DbState dbState = StateGeneratorUtils.generateDbState(getPairToCursorInfoMap())
        .withCdc(true)
        .withCdcState(getCdcStateManager().getCdcState());

    return new AirbyteStateMessage()
        .withStateType(AirbyteStateType.GLOBAL)
        // Temporarily include legacy state for backwards compatibility with the platform
        .withData(Jsons.jsonNode(dbState))
        .withGlobal(globalState);
  }
```
But when we read it back we are trying to map it into the `DbState` class which seems incorrect. I think the usage of DbState class should only be for Legacy i.e. for data field only and once the conversion has happened we should not try to map things to DbState cause the state has been split in shared and individual streams

Reply:
> Here is my concern with this method, when we emit the state we split the JSON into `sharedState` and `streams`
> 
> ```
>   @Override
>   public AirbyteStateMessage toState(final AirbyteStreamNameNamespacePair pair) {
>     // Populate global state
>     final AirbyteGlobalState globalState = new AirbyteGlobalState();
>     globalState.setSharedState(Jsons.jsonNode(getCdcStateManager().getCdcState()));
>     globalState.setStreamStates(StateGeneratorUtils.generateStreamStateList(getPairToCursorInfoMap()));
> 
>     // Generate the legacy state for backwards compatibility
>     final DbState dbState = StateGeneratorUtils.generateDbState(getPairToCursorInfoMap())
>         .withCdc(true)
>         .withCdcState(getCdcStateManager().getCdcState());
> 
>     return new AirbyteStateMessage()
>         .withStateType(AirbyteStateType.GLOBAL)
>         // Temporarily include legacy state for backwards compatibility with the platform
>         .withData(Jsons.jsonNode(dbState))
>         .withGlobal(globalState);
>   }
> ```
> 
> But when we read it back we are trying to map it into the `DbState` class which seems incorrect. I think the usage of DbState class should only be for Legacy i.e. for data field only and once the conversion has happened we should not try to map things to DbState cause the state has been split in shared and individual streams

@subodh1810 This is here for backwards compatibility/to work with the scenario where the connector has this code, but the platform doesn't.